### PR TITLE
speedup: switch EgammaTools EnergyScaleCorrection scales and smearings to use std::map (backport of #32053)

### PR DESCRIPTION
**Pull request by slava77**
backport of #32053

this is a technical PR to speedup the startup time of workflows using `CalibratedPhotonProducer` and `CalibratedElectronProducer` which use `EnergyScaleCorrection`

more details are available in https://github.com/cms-sw/cmssw/pull/32053#issue-517027691

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoEgamma/EgammaTools/interface/EnergyScaleCorrection.h [hunk 1]
```diff
@@ -193,8 +194,8 @@ class EnergyScaleCorrection {
 
   //data members
   FileFormat smearingType_;
-  std::vector<std::pair<CorrectionCategory, ScaleCorrection> > scales_;
-  std::vector<std::pair<CorrectionCategory, SmearCorrection> > smearings_;
+  std::map<CorrectionCategory, ScaleCorrection> scales_;
+  std::map<CorrectionCategory, SmearCorrection> smearings_;
```
Comment (VinInn) [inline RecoEgamma/EgammaTools/interface/EnergyScaleCorrection.h:198]: why not unordered_map?

Comment (slava77) [reply RecoEgamma/EgammaTools/interface/EnergyScaleCorrection.h:198]: This idea crossed my mind.
However, it looked like some of the logic implied that the key ordering is important.
I didn't make enough effort to prove this wrong to myself.

Comment (slava77) [reply RecoEgamma/EgammaTools/interface/EnergyScaleCorrection.h:198]: Oh, also, I'm was not sure how to write a hash that would return the right value starting from `{2 ints, 3 doubles}`, which would correspond to a range of values mapping on the same key.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/32055	124cd3c 
- the updates are identical to the master version (#32053 ), which was merged a few days ago
- signing off after the additional OK in https://github.com/cms-sw/cmssw/pull/32053#issuecomment-724825163

**Comment by qliphy** [comment]
+1

**Pull request merged**